### PR TITLE
Follow up code health fixes of PR 1442

### LINKED ISSUE #1040: Add ability to validate the min/max for Date on a FHIR Questionnaire
**Is your feature request related to a problem? Please describe.**
Add ability to limit the date input values to either the minimum or maximum date allowed for a given use case e.g. 
1. A task should not start before a certain date 
2. When adding DOB, the DOB should be less than the current day/today 
3. When scheduling you can validate when a given task should be done or assigned to a given practitioner 

**Describe the solution you'd like**
1. To be able to resolve expressions such as today() in the min-max value extension
2. To use the min-max value extension in the actual date picker pop-up. at the moment the pop-up dialogue allows the user to put in any date and the error message only kicked in once the user has input a date.

**Describe alternatives you've considered**
1.  It would be much better if the date picker already doesn't allow the user to input a date outside the allowed range.

**Additional context**
A detailed use case is defined on this ticket https://github.com/opensrp/fhircore/issues/845 on OpenSRP FHIR Core repo

**Would you like to work on the issue?**
@RaaziaTarique

### PR DESCRIPTION
**Pull request by aurangzaibumer**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1040 

**Description**
The min/max validators are updated to use fhir path expressions. For date picker specifically, the widget doesn't allow input outside of the date range, this is a widget specific behavior. User can specify minimum and maximum date range using minValue and maxValue expression in questionnaire in order to prevent selecting the dates from outside the specified range. cqf-calculatedvalue expression is used to make the expression dynamic and use for some purpose as Age case (for example we can only registered a child that is born after 2020) (min value case).
 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
if this is ready for reveiw can you please mark this as such?

thanks - also, probably best for @omarismail94 to review since he reviewred the original pr.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 1]
```diff
@@ -54,7 +54,7 @@ object ExpressionEvaluator {
    */
   private val variableRegex = Regex("[%]([A-Za-z0-9\\-]{1,64})")
 
-  private val fhirPathEngine: FHIRPathEngine =
+  internal val fhirPathEngine: FHIRPathEngine =
```
Comment (aurangzaibumer) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:57]: https://github.com/google/android-fhir/pull/1442#discussion_r975439051

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt [hunk 2]
```diff
@@ -16,27 +16,20 @@
...
+internal const val CQF_CALCULATED_EXPRESSION_URL: String =
+  "http://hl7.org/fhir/StructureDefinition/cqf-calculatedValue"
 
 fun Type.valueOrCalculateValue(): Type? {
   return if (this.hasExtension()) {
     this.extension.firstOrNull { it.url == CQF_CALCULATED_EXPRESSION_URL }?.let {
       val expression = (it.value as Expression).expression
-      fhirPathEngine.evaluate(this, expression).firstOrNull()?.let { it as Type }
+      fhirPathEngine.evaluate(this, expression).singleOrNull()?.let { it as Type }
     }
```
Comment (aurangzaibumer) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt:31]: https://github.com/google/android-fhir/pull/1442#discussion_r975440215

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/ValidationUtilTest.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.hl7.fhir.r4.model.Expression
+import org.hl7.fhir.r4.model.Extension
+import org.hl7.fhir.r4.model.IntegerType
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@RunWith(RobolectricTestRunner::class)
```
Comment (aurangzaibumer) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/ValidationUtilTest.kt:34]: https://github.com/google/android-fhir/pull/1442#discussion_r975442029

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidator.kt [hunk 4]
```diff
@@ -25,6 +25,7 @@ import org.hl7.fhir.r4.model.QuestionnaireResponse
 import org.hl7.fhir.r4.model.Type
 
 internal const val MAX_VALUE_EXTENSION_URL = "http://hl7.org/fhir/StructureDefinition/maxValue"
```
Comment (aurangzaibumer) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidator.kt:27]: https://github.com/google/android-fhir/pull/1442#discussion_r975444895

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidatorTest.kt [hunk 5]
```diff
@@ -200,14 +140,14 @@ class MaxValueConstraintValidatorTest {
         }
       )
 
-    assertTrue(
-      (MaxValueConstraintValidator.getMaxValue(questionItem.first()) as? DateType)?.valueAsString
-        ?.equals(today) == true
-    )
+    assertThat(
```
Comment (aurangzaibumer) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidatorTest.kt:143]: https://github.com/google/android-fhir/pull/1442#discussion_r975509975

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt [hunk 6]
```diff
@@ -50,7 +50,7 @@ class MinValueConstraintValidatorTest {
   }
 
   @Test
-  fun shouldReturnInvalidResult() {
+  fun `should return invalid result`() {
```
Comment (maimoonak) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt:53]: Plz change test names to be a little more descriptive.. why it should return invalid result

**Review by maimoonak — COMMENTED** [review]
Plz make test names a little more descriptive ... should like `should return invalid value`... why it should return invalid value?

**Comment by aurangzaibumer** [comment]
Comments are resolved. Please review.

**Review by maimoonak — APPROVED** [review]

**Comment by aurangzaibumer** [comment]
Review completed by @maimoonak . Assigned to @omarismail94 as per review process. thanks
cc @f-odhiambo

**Review by jingtang10 — APPROVED** [review]
thanks @aurangzaibumer

**Pull request merged**